[Android] Fix Border RoundRectangle CornerRadius issue

jsuarezruiz opened the pull request:
### Description of Change

Fix **Border** RoundRectangle CornerRadius issue on Android.

Fixes #14936, same corner radius in all the platforms. In the following capture, the top Border is on Windows and the bottom one on Android.
![fix-corner](https://github.com/dotnet/maui/assets/6755973/bb2600e3-03f1-4026-b28a-a4228f54c6f0)

Fixes #14276, on the right side the Android version with the issue:
![image](https://github.com/dotnet/maui/assets/6755973/c1fbb8dc-464d-496c-8e02-21c78e4edfab)
With the changes:
![fix-14276](https://github.com/dotnet/maui/assets/6755973/c4b878b7-27ab-44fb-bb75-f68de2cc01e9)

### Issues Fixed

Fixes #14936
Fixes #14276

jsuarezruiz committed:
Fix Droid Border RoundRectangle Corner Radius issue

Commit:
Auto-format source code

jsuarezruiz commented on src/Controls/src/Core/HandlerImpl/RoundRectangle.Impl.cs:
```diff
@@ -51,10 +52,14 @@ public override PathF GetPath()
-			float bottomRightCornerRadius = (float)CornerRadius.BottomRight;
+			float density = 1f;
+#if ANDROID
+			density = (float)DeviceDisplay.MainDisplayInfo.Density;
```
Reviewing why we need this changes and if have more sense to apply it at other level. We have not used density before.

jsuarezruiz replied:
Maybe related to https://github.com/dotnet/maui/commit/982d809eb9567f75ea810a3d7aedb6aac8f8284c#diff-9962b2ab4d4a0eb4922307e668bfe7d868c4fac7919a9463a3cba1859a6de86fL417-L423

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Moved changes

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

jsuarezruiz committed:
Merge branch 'main' into fix-14936

Redth commented on src/Controls/src/Core/Shapes/RoundRectangle.cs:
```diff
@@ -66,10 +67,14 @@ public override PathF GetPath()
-			float bottomRightCornerRadius = (float)CornerRadius.BottomRight;
+			float density = 1f;
+#if ANDROID
+			density = (float)DeviceDisplay.MainDisplayInfo.Density;
```
```suggestion
			density = Handler?.MauiContext?.Context?.GetDisplayDensity()
	?? (float)DeviceDisplay.MainDisplayInfo.Density;
```

Redth commented on src/Controls/src/Core/Shapes/RoundRectangle.cs:
```diff
@@ -101,10 +106,14 @@ internal PathF GetInnerPath(float strokeWidth)
-			float bottomRightCornerRadius = (float)Math.Max(0, CornerRadius.BottomRight - strokeWidth);
+			float density = 1f;
+#if ANDROID
+			density = (float)DeviceDisplay.MainDisplayInfo.Density;
```
```suggestion
			density = Handler?.MauiContext?.Context?.GetDisplayDensity()
	?? (float)DeviceDisplay.MainDisplayInfo.Density;
```

jstedfast committed:
Improved the RoundRectangle Border device test for better accuracy

jstedfast committed:
De-duplicate logic for getting the Inner and Outer Paths for RoundRectangle

jstedfast committed:
Removed unused using statement

jstedfast committed:
Merge branch 'main' into fix-14936

jstedfast committed:
Fix ROundRectangle Border device test for ANdroid

(and hopefully Windows/iOS)

jstedfast commented on src/TestUtils/src/DeviceTests/AssertionExtensions.Windows.cs:
```diff
@@ -394,6 +394,18 @@ public static async Task<CanvasBitmap> AssertColorAtPointAsync(this FrameworkEle
 
+		public static async Task<CanvasBitmap> AssertColorsAtPointsAsync(this FrameworkElement view, Graphics.Color[] colors, Graphics.Point[] points, IMauiContext mauiContext)
+		{
+			var bitmap = await view.ToBitmap(mauiContext);
```
yes, it is intentionally like all of the other helper methods that do the same.

jstedfast committed:
Try adding a slight adjustment to point locations for Border tests

jstedfast commented:
At this point I just need to adjust the Points in the DeviceTests to work for iOS.

jstedfast committed:
Potential fix for iOS RoundRectangle BorderTests by using scale factor

jstedfast commented:
Looks like the iOS Border tests are passing now, but oddly it's getting other errors.

jstedfast committed:
Disable the iOS RoundRectangle test - seems to draw w/o any border??

Redth committed:
Use CreateColorAtPointError to capture image

This captures the image of what the assertion looked like and dumps it out as base64 where the test runner can pick it up and actually display it.

jstedfast committed:
Fixed iOS RoundRectangle BorderTest offsets

jstedfast commented:
Seems like random tests are failing

jstedfast committed:
Merge branch 'main' into fix-14936

PureWeen reviewed (dismissed):
Not sure if this is related to this PR but I noticed in android the border sizing doesn't seem correct. 

I merged in your iOS border samples and notice that android has an odd black border between the red and blue. 

![image](https://github.com/dotnet/maui/assets/5375137/c18fb1bc-6803-42d5-aa53-438e833a8d83)

rmarinho committed:
Merge branch 'main' into fix-14936

jstedfast committed:
Merge branch 'fix-14936' of https://github.com/dotnet/maui into fix-14936

jstedfast committed:
Merge branch 'main' into fix-14936

jstedfast committed:
Update HandlerTestBasement.cs

jstedfast committed:
Merge branch 'main' into fix-14936

jstedfast commented:
I'm beginning to wonder if the density calculation is wrong. It seems like MauiDrawable.Android.cs should pass in `_strokeThickness / _density` instead of `_strokeThickness` because `strokeThickness` has already been multiplied by `_density` but all of the path calculations assume "Maui coordinates" instead of "Android coordinates".

jstedfast commented:
Replace by https://github.com/dotnet/maui/pull/17087
